### PR TITLE
Mongoose session

### PR DESCRIPTION
**Pull request by NelsonVides**
Plenty of explanations and profiling and details and whatnot, in the commit messages 😁

### CODE REVIEW

**Comment by mongoose-im** [comment]
[9049.1](https://travis-ci.org/esl/MongooseIM/jobs/756765213) / Erlang 23.0.3 / small_tests / 8c2297fbfe8de6703d5af97a9c6bf94d8ff44e93
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/3018/9049/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/3018/9049/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2021-01-29_22.53.53/lib.mongooseim.logs/run.2021-01-29_22.53.54/suite.log.html)

File: src/mod_carboncopy.erl [hunk 1]
```diff
@@ -300,29 +300,33 @@ complete_packet(_From, #xmlel{name = <<"message">>, attrs=OrigAttrs} = Packet, r
...
+              end
+      end,
+      AllSessions).
+
+filter_priority_resources(AllSessions) ->
+    lists:filtermap(
+      fun(#session{usr = {_, _, R}, priority = P}) when is_integer(P) -> {true, {P, R}};
+         (_) -> false
+      end,
+     AllSessions).
```
Comment (chrzaszcz) [inline src/mod_carboncopy.erl:326]: Minor, not a must: here I'd rather have the inner `fun` as a separate function and call it like `lists:filtermap(fun filter_cc_enabled_resource/1, AllSessions)`. Reason: to have function names for tracing and debugging as the funs are not trivial.

File: src/ejabberd_sm.erl [hunk 2]
```diff
@@ -128,7 +128,7 @@
                       priority :: priority(),
                       info     :: info()
                      }.
--type info() :: [info_item()].
+-type info() :: #{info_key() => any()}.
 
 %% Session representation as 4-tuple.
 -type ses_tuple() :: {USR :: jid:simple_jid(),
```
Comment (chrzaszcz) [inline src/ejabberd_sm.erl:134]: It looks like we are leaving an unused type behind, could you validate that nothing else is missed? I only checked this one.

**Review by chrzaszcz — COMMENTED** [review]
Good changes! A few comments from me.

File: src/ejabberd_sm_redis.erl [hunk 3]
```diff
@@ -34,32 +34,24 @@ start(_Opts) ->
...
     Keys = mongoose_redis:cmd(["KEYS", hash(<<"*">>)]),
     lists:flatmap(fun(K) ->
                           Sessions = mongoose_redis:cmd(["SMEMBERS", K]),
                           lists:map(fun(S) ->
-                                            Session = binary_to_term(S),
-                                            { Session#session.usr,
-                                              Session#session.sid,
-                                              Session#session.priority,
-                                              Session#session.info }
+                                            binary_to_term(S)
```
Comment (chrzaszcz) [inline src/ejabberd_sm_redis.erl:43]: I wonder how performant this is (mem, CPU) compared to tuples.

Comment (NelsonVides) [reply src/ejabberd_sm_redis.erl:43]: `binary_to_term` returns a C array of size six with six pointers to six created-on-the-stop ERL_TERMs.
Then the next instruction allocates a new C array of size four, and copies four specific fields of those pointers from the previous array. It shouldn't need to copy the contents of the pointers, as they're immutable. So technically speaking all that is happening is an extra 4*64bit memory allocation, with the CPU it takes to allocate that and later on GC both arrays. Difference is probably so so minimal it really doesn't matter, but it's still totally unnecessary 🤷🏽‍♂️ Also, records are clearer than arbitrary tuples 😐

File: src/mod_carboncopy.erl (hunk 1 above)
Comment (NelsonVides) [reply src/mod_carboncopy.erl:326]: Sure thing, good argument 👍🏽

File: src/ejabberd_sm_redis.erl (hunk 3 above)
Comment (chrzaszcz) [reply src/ejabberd_sm_redis.erl:43]: Never mind, I thought for some reason that this PR changed the data stored in Redis, which is not the case.

**Review by chrzaszcz — APPROVED** [review]
Very good changes!

**Pull request merged**